designs/DeveloperTestingPackageExperience.md

```diff
@@ -0,0 +1,125 @@
+package may be an existing published version or that of a previously restored
+local package (i.e. there is no requirement that the version is unique, as there
+is with the current package publishing requirements), however, it is assumed
+that the package ID will be unique to avoid conflict with other dependencies.
+
+## Non-Goals
+
+Custom feeds will not be supported by the solution - the solution is only
+interested in folders containing `nupkg` files. The rationale for this is to
+keep the design simple and to emphasise that it is not a replacement for other
+package sources. Also, since the target audience is creating the packages, the
+file would have been produced in order to publish it, so this limitation is not
+perceived as an impediment.
+
+## Solution Overview
+
+In order to provide a solution, there are a few issues that need to be
+highlighted with the current machinery with regards to installing of a package
+in the context of trying to test a newly created package:
+
+* NuGet caches packages by ID and version
+* The order of multiple sources are searched through is non-deterministic
+* If a version number is omitted, NuGet searches for the lowest applicable
+  version
+
+### `csproj` Additions
+
+The proposed solution is to add an additional MsBuild property for the
+[restore target](https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#restore-target)
+as follows:
+
+| Property              | Description  |
+|-----------------------|--------------|
+| RestoreOverrideFolder | Specifies a folder that contains `nupkg` files that take precedence over any other source and are not cached. |
+
+This property can only be specified once and can only specify a single folder
+(i.e. there cannot be multiple override folders to avoid ordering issues). Note
+that file shares _should_ be supported, however.
+
+Also, only a single version of a package (based on its ID) may exist in the
+override folder (i.e. it would be an error to have
+`generated.package.1.0.0.nupkg` and `generated.package.2.0.0.nupkg` in the
+folder).
+
+### Resolving Changes
+
+When an override folder is specified, any packages found inside the folder
+should be removed from the current cache (matching by ID and version) before any
+further work is done (i.e. this would delete the folder from the global cache
+or from the `PACKAGES_DIRECTORY` if specified in the
+[restore options](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-restore#options)).
+
+When a package that exists inside the overrides folder is referenced by the
+project:
+
+* If no version is specified, the version in the folder is used without
+  searching any other sources.
+* If a version is specified and that matches that in the override folder (taking
+  into account the usual wildcard support), that package is used without
+  searching any other sources.
+* Otherwise, normal resolution happens.
+
+When the override package is used, it is restored to the cache as normal (taking
```

Comment:
Note that `no-cache` has nothing to do with the global packages folder, which is an installation directory. 
It refers to the http-cache.

Reply:
Here I was trying to refere to the packages folder (either global or where specifed by the `--packages` option). Is there a better term for this (packages folder perhaps, though that could be the taken as the source packages folder)?

Reply:
When you specify `--packages` you are overriding the global packages folder value, so I think using global packages folder and packages folder in this context is good.